docs/development/vscode.md

```diff
@@ -0,0 +1,143 @@
+    ```
+
+6. Expose port in Docker
+
+    If you are using a *virtual box* instance, you can skip this step.
+    Add the following lines at the end of the
+    `tools/setup/dev-vagrant-docker` and replace
+    `YOUR_PTVSD_PORT_NUMBER` with the port number on which ptvsd server
+    will run. (This must be same as the `YOUR_HOST_PORT_NUMBER` from
+    last step)
+    ```docker
+    EXPOSE YOUR_PTVSD_PORT_NUMBER
+    ```
+    Example:
+
+    ```docker
+    EXPOSE 5678
+    ```
+
+7. Download and install ptvsd by `ssh-ing` into the vagrant server.
+
+    **Note**: You should activate the `zulip-py3-venv` if not already
+    activated and then run the following.
+
+    ```python
+    pip install ptvsd
+    ```
+
+8. Edit `tools/settings.py`
+
+    Add the following lines in `zproject/settings.py` after `if DEBUG:`
+    ```python
+    if DEBUG:                                   # Already present
+        INTERNAL_IPS = ('127.0.0.1',)           # Already Present
+        import ptvsd
+        if "runserver" in sys.argv and os.environ.get("RUN_MAIN"):
+            try:
+                ptvsd.enable_attach(redirect_output=True)
+                print("VS Code Debugging Enabled.")
+            except OSError:
+                print("Debugging Setup Error, Exception: ")
+                tracback.print_exc()
+    ```
+
+    The `redirect_output` gets the output from the server redirected to
+    the debug console of VS Code (quite a handy thing!).
+
+    Here you can have a bit of modifications like:
+    - If you want the django server to wait for debugger *prior to
+      running* just add
+    ```python
+    ...
+    ptvsd.enable_attach()
+    ptvsd.wait_for_debugger()
+    ...
+    ```
+   - If you want to run the ptvsd server on some other port than default
+     (default is *0.0.0.0* and *5678* ) then
+    ```python
+    ptvsd.enable_attach(YOUR_HOST_IP, YOUR_HOST_PORT)
+    ```
+
+9. Create a `launch.json` file
+
+    Add a debugging config by checking out the instructions
+    [here](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations).
+    In the launch.json add the following lines
+
+    ```js
+    {
+        "version": "0.2.0",
+        "configurations": [
+            {
+            "name":"Python: Remote Attach",     // name of config
+            "type":"python",
+            "request":"attach",                 // attach type request
+            "port":5678,                        // the port on which ptvsd is running
+            "host":"localhost",                 // host ip of the vagrant server
+            "django" : true,                    // Not neccessary
+            "subprocess" : true,                // multi-process architecture
+            "pathMappings":[{                   // your local->server mappinps
+                "localRoot":"${workspaceFolder}",
+                "remoteRoot":"."
+                }]
+            }  
+        ]
+    }
+    ```
```

Comment:
Could we commit this file as `tools/setup/vscode-launch.json` to make it easy for folks to just copy the file?

Reply:
We will just keep the barebones required for debugging with extra options left with the developer to make the adjustments. The extra options can be found [here](https://github.com/microsoft/ptvsd) as well as [here](https://code.visualstudio.com/docs/python/debugging). 

Maybe we should highlight this in the `tools/setup/vscode-launch.json` as a comment for the developer.